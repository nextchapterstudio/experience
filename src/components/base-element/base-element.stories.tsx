import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import { BaseElement } from './base-element'

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'Example/BaseElement',
  component: BaseElement,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
    backgroundColor: { control: 'color' },
  },
} as ComponentMeta<typeof BaseElement>

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof BaseElement> = (args) => <BaseElement {...args} />

export const Primary = Template.bind({})
// More on args: https://storybook.js.org/docs/react/writing-stories/args
Primary.args = {
  children: 'BaseElement Text',
}

export const AsASpan = Template.bind({})
AsASpan.args = {
  children: 'BaseElement Text',
  as: 'span',
}
