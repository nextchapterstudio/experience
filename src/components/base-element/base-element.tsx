import React, { forwardRef, ElementType, HTMLAttributes, FC, ComponentProps  } from 'react'
import classnames from 'classnames'

import { allTheClasses } from '../../utils/modifiers'
import htmlAttributes, { ariaAttributes } from '../../utils/element-attributes'
import { propTypes } from '../../utils/prop-types'

export const BaseElement: FC<BaseProps> = forwardRef(({ as: Element = 'div', className, children, ...otherProps }, ref) => {

    const classes = allTheClasses(otherProps)
    const { onClick } = otherProps
    const tagName =Element?.toString() || 'div'


    const castedProps = otherProps as Record<string, string>
    const clean = htmlAttributes['*']
      .concat(htmlAttributes[tagName] || [])
      .concat(ariaAttributes)
      .reduce((objs: Record<string, string>, current: string) => {
        if (castedProps[current]) {
          objs[current] = castedProps[current]
        }
        return objs
      }, {})

    return (
      <Element
        ref={ref}
        className={classnames(className, classes)}
        onClick={onClick}
        {...clean}
      >
        {children}
      </Element>
    )
  })


export interface BaseProps extends HTMLAttributes<HTMLOrSVGElement>  {
    as?: ElementType;
    children?: React.ReactNode
    className?: string
    isSpaced?: boolean
    status?: 'inactive' | 'inProgress' | 'success' | 'failed'
    value?: string | number | Error
    onClick?: () => void;
  }


BaseElement.defaultProps = {
  as: 'div',
}
