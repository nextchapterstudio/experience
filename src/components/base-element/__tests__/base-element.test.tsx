import React from 'react'
import { render, cleanup, fireEvent, screen } from '@testing-library/react'
import { BaseElement } from '../base-element'

describe('The Base Element Component', () => {
  afterEach(cleanup)
  it('should render a div with the class basic', () => {
    const { container } = render(<BaseElement className='basic'>test</BaseElement>)

    expect(container).toMatchSnapshot()
    expect(container.firstChild).toMatchSnapshot()
    expect(container.firstChild).toHaveClass('basic')
  })

  it('should render a span with the class basic', () => {
    const { container } = render(<BaseElement as="span" className='basic' >test</BaseElement>)

    expect(container).toMatchSnapshot()
    expect(container.firstChild).toMatchSnapshot()
    expect(container.firstChild).toHaveClass('basic')
  })
  it('should render a div with the html attribute', () => {
    const { container } = render(<BaseElement className='basic' aria-hidden>test</BaseElement>)

    expect(container).toMatchSnapshot()
    expect(container.firstChild).toMatchSnapshot()
    expect(container.firstChild).toHaveAttribute('aria-hidden')
  })
  it('should render as a p tag', () => {
    const { container } = render(<BaseElement className='basic' as="p">test</BaseElement>)
    expect(container).toMatchSnapshot()
    expect(container).toContainHTML('<p')
  })
  it('should add an onclick', async () => {
    let value = 0
    const { container } = render(<BaseElement className='basic' onClick={()=> {
      value = 2
    }}>test</BaseElement>)
    fireEvent.click(await screen.findByText('test'))
    expect(container).toMatchSnapshot()
    expect(container).toContainHTML('<p')
    expect(value).toBe(2)
  })
})
