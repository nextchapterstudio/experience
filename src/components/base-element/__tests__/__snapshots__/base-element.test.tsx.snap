// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Base Element Component should add an onclick 1`] = `
<div>
  <div
    class="basic"
  >
    test
  </div>
</div>
`;

exports[`The Base Element Component should render a div with the class basic 1`] = `
<div>
  <div
    class="basic"
  >
    test
  </div>
</div>
`;

exports[`The Base Element Component should render a div with the class basic 2`] = `
<div
  class="basic"
>
  test
</div>
`;

exports[`The Base Element Component should render a div with the html attribute 1`] = `
<div>
  <div
    aria-hidden="true"
    class="basic"
  >
    test
  </div>
</div>
`;

exports[`The Base Element Component should render a div with the html attribute 2`] = `
<div
  aria-hidden="true"
  class="basic"
>
  test
</div>
`;

exports[`The Base Element Component should render a span with the class basic 1`] = `
<div>
  <span
    class="basic"
  >
    test
  </span>
</div>
`;

exports[`The Base Element Component should render a span with the class basic 2`] = `
<span
  class="basic"
>
  test
</span>
`;

exports[`The Base Element Component should render as a p tag 1`] = `
<div>
  <p
    class="basic"
  >
    test
  </p>
</div>
`;
